tooling/metamodel-generator/src/main/java/org/hibernate/processor/ImportContextImpl.java

```diff
@@ -118,12 +119,11 @@ else if ( result.endsWith( "..." ) ) {
 			result = result.substring( 0, index );
 		}
 
-		return ( preamble + unqualifyName( result ) + appendices )
-				.replace( '$', '.' );
+		return removeDollar( (preamble + unqualifyName( result ) + appendices) );
 	}
 
 	private String unqualifyName(String qualifiedName) {
-		final String sourceQualifiedName = qualifiedName.replace( '$', '.' );
+		final String sourceQualifiedName = removeDollar( qualifiedName );
```

Comment:
This change looks wrong to me. `removeDollar()` removes a trailing dollar from a class name generated by Scott's tool for implementing query-by-method-name.

What we're trying to do here is replace a `$` in the _middle of_ a name of a class file corresponding to an inner class with a dot to get its name in the source code.

Reply:
Current code will change that trailing dollar to dot, and that is really bad thing :-)

Other solution can be something like 
```
qualifiedName.substring( 0, qualifiedName.length() - 1 ).replace( '$', '.' ) + qualifiedName.charAt( qualifiedName.length() - 1 )
```

Not nice looking thing, but this will change all `$`'s into dots, except one at last position.

Reply:
> Current code will change that trailing dollar to dot, and that is really bad thing :-)

Ah well I guess that's because `TypeUtils.getGeneratedClassFullyQualifiedName()` should not be calling `removeDollar()`.

Reply:
One solution is to skip `importContext.importType( element.getQualifiedName().toString() )` if qualified (or simple name) is ending with `$`. It will not be used in generated meta model class, anyway. This way only change from `main`will be in `AnnotationMetaEntity`:

```
		if ( !element.getQualifiedName().toString().endsWith( "$" ) ) {
			importContext.importType( element.getQualifiedName().toString() );
		}
```

This will work fine with all existing tests, but will show that there is problem with dollar-ending inner class. This can be solved by not creating meta model class for injectable annotation meta entity that is not initialized (one not ending with dollar)

Reply:
I'm not really sure what problem you're trying to solve here. The class names with trailing `$`s are just a special case for the query by method name stuff. We're not expecting to encounter them in user code. I'm not even sure why `getGeneratedClassFullyQualifiedName()` calls `removeDollar()`, I don't think it should.

Reply:
>  I'm not even sure why `getGeneratedClassFullyQualifiedName()` calls `removeDollar()`, I don't think it should.

Excuse me, I'm wrong, it _is_ needed, because the test for that stuff fails without it.